tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+) -> Option<Waker> {
+    loop {
+        match curr {
+            EMPTY | NOTIFIED => {
+                let res = state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return None,
+                    Err(actual) => {
+                        curr = actual;
+                    }
+                }
+            }
+            WAITING => {
+                // At this point, it is guaranteed that the state will not
+                // concurrently change as holding the lock is required to
+                // transition **out** of `WAITING`.
+                //
+                // Get a pending waiter
+                let mut waiter = waiters.pop_back().unwrap();
+
+                assert!(!waiter.notified);
+
+                waiter.notified = true;
+                let waker = waiter.waker.take();
+
+                if waiters.is_empty() {
+                    // As this the **final** waiter in the list, the state
+                    // must be transitioned to `EMPTY`. As transitioning
+                    // **from** `WAITING` requires the lock to be held, a
+                    // `store` is sufficient.
+                    state.store(EMPTY, SeqCst);
+                }
+
+                return waker;
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+// ===== impl RecvFuture =====
+
+impl RecvFuture<'_> {
+    fn project(
+        self: Pin<&mut Self>,
+    ) -> (
+        &Notify,
+        &mut RecvState,
+        Pin<&mut linked_list::Entry<Waiter>>,
+    ) {
+        unsafe {
+            // Safety: both `notify` and `state` are `Unpin`.
+
+            is_unpin::<&Notify>();
+            is_unpin::<AtomicU8>();
+
+            let me = self.get_unchecked_mut();
+            (
+                &me.notify,
+                &mut me.state,
+                Pin::new_unchecked(&mut me.waiter),
+            )
+        }
+    }
+}
+
+impl Future for RecvFuture<'_> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        use RecvState::*;
+
+        let (notify, state, mut waiter) = self.project();
+
+        'outer: loop {
+            match *state {
+                Init => {
+                    // Optimistically try acquiring a pending notification
+                    let res = notify
+                        .state
+                        .compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst);
+
+                    if res.is_ok() {
+                        // Acquired the notification
+                        *state = Done;
+                        continue 'outer;
+                    }
+
+                    // Acquire the lock and attempt to transition to the waiting
+                    // state.
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    // Reload the state with the lock held
+                    let mut curr = notify.state.load(SeqCst);
```

Comment:
Have you thought about just storing the `Waker` in the list, unlock it, and then check `state` again and act on a missed notification instead of trying to reason about the Mutex and the atomic at the same time?

I have the feeling it might be slightly less efficient for the case where the race actually happens, but might be easier to follow.

Reply:
It might work... not sure. I  mostly  modeled it after the park / unpark impl from `std`. Right now, it is roughly "check fast path, otherwise, lock and do the full work". This feels simpler than trying to "undo" work, but I can't say for sure.

 If you think it would be worth it, I can give a shot. wdyt?

Reply:
I have no strict preference on this one. Whatever seems to work and is understood by reviewers is fine. I thought an approach where you retry might be easier because of less require extra synchronization and code paths. The code for checking for a notification is already there, and you might just need to use it again. It's kind of the compare-and-swap approach we generally use for atomics.

Reply:
I will opt to leave it this way as it follows the style of `park` / `unpark`... and i prefer not changing code if it isn't requested 😆